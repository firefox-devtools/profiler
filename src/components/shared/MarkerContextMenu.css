/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

/* Most of the CSS logic to be able to add icons is part of the more generic
 * file ContextMenu.css that is applied for all context menus. Especially the
 * CSS class `react-contextmenu-icon` is used in addition to these more specific
 * classes, and that's where other properties for backgrounds are specified.
 */

/* Here are all icons for the marker context menu. */
.markerContextMenuIconSetSelectionFromMarker {
  background-image: url(../../../res/img/svg/set-selection-from-marker.svg);
}

.markerContextMenuIconStartSelectionHere, /* <- for dot markers */
.markerContextMenuIconStartSelectionAtMarkerStart /* <- for interval markers */ {
  background-image: url(../../../res/img/svg/start-selection-at-marker-start.svg);
}

.markerContextMenuIconStartSelectionAtMarkerEnd {
  background-image: url(../../../res/img/svg/start-selection-at-marker-end.svg);
}

.markerContextMenuIconEndSelectionAtMarkerStart {
  background-image: url(../../../res/img/svg/end-selection-at-marker-start.svg);
}

.markerContextMenuIconEndSelectionHere, /* <- for dot markers */
.markerContextMenuIconEndSelectionAtMarkerEnd /* <- for interval markers */ {
  background-image: url(../../../res/img/svg/end-selection-at-marker-end.svg);
}

.markerContextMenuIconCopyDescription {
  background-image: url(../../../res/img/svg/copy-dark.svg);
}

.markerContextMenuIconCopyStack {
  background-image: url(../../../res/img/svg/copy-marker-stack.svg);
}

.markerContextMenuIconCopyUrl {
  background-image: url(../../../res/img/svg/globe-dark.svg);
}

.markerContextMenuIconCopyPayload {
  background-image: url(../../../res/img/svg/copy-marker-payload.svg);
}

.markerContextMenuSelectThread {
  background-image: url(../../../res/img/svg/select-thread.svg);
}

/* These icons don't look nice when selected, unless if we invert them. However
 * the other icons look better without the filter, so we don't change them in
 * this state. */
.react-contextmenu-item--selected .markerContextMenuIconCopyDescription,
.react-contextmenu-item:hover .markerContextMenuIconCopyDescription,
.react-contextmenu-item--selected .markerContextMenuIconCopyUrl,
.react-contextmenu-item:hover .markerContextMenuIconCopyUrl,
.react-contextmenu-item--selected .markerContextMenuIconCopyStack,
.react-contextmenu-item:hover .markerContextMenuIconCopyStack,
.react-contextmenu-item--selected .markerContextMenuIconCopyPayload,
.react-contextmenu-item:hover .markerContextMenuIconCopyPayload,
.react-contextmenu-item--selected .markerContextMenuSelectThread,
.react-contextmenu-item:hover .markerContextMenuSelectThread {
  filter: invert(1);
}
