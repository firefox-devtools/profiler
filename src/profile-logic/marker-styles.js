/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
// @flow
import * as colors from 'photon-colors';

import type { CssPixels, Marker } from 'firefox-profiler/types';

type MarkerStyle = {
  +top: CssPixels,
  +height: CssPixels,
  +background: string,
  +squareCorners: boolean,
  +borderLeft: null | string,
  +borderRight: null | string,
};

const defaultStyle = {
  top: 0,
  height: 6,
  background: 'black',
  squareCorners: false,
  borderLeft: null,
  borderRight: null,
};

const gcStyle = {
  ...defaultStyle,
  top: 6,
  background: colors.ORANGE_50,
};

const ccStyle = {
  ...gcStyle,
  // This is a paler orange to distinguish CC from GC.
  background: '#ffc600',
};

/**
 * Get the marker style. Start off by looking at the marker name, then fallback to
 * the marker type.
 */
export function getMarkerStyle(marker: Marker): MarkerStyle {
  const { data, name } = marker;
  if (name in markerStyles) {
    return markerStyles[name];
  }
  if (data && data.type in markerStyles) {
    return markerStyles[data.type];
  }
  return markerStyles.default;
}

const markerStyles: { +[styleName: string]: MarkerStyle } = {
  default: defaultStyle,
  RefreshDriverTick: {
    ...defaultStyle,
    background: 'hsla(0,0%,0%,0.05)',
    height: 18,
    squareCorners: true,
  },
  RD: {
    ...defaultStyle,
    background: 'hsla(0,0%,0%,0.05)',
    height: 18,
    squareCorners: true,
  },
  // Scripts is renamed to 'requestAnimationFrame callbacks' but keeping this
  // here for backwards compatibility.
  Scripts: {
    ...defaultStyle,
    background: colors.ORANGE_70,
    top: 6,
  },
  'requestAnimationFrame callbacks': {
    ...defaultStyle,
    background: colors.ORANGE_70,
    top: 6,
  },
  Styles: {
    ...defaultStyle,
    background: colors.TEAL_50,
    top: 7,
  },
  FireScrollEvent: {
    ...defaultStyle,
    background: colors.ORANGE_70,
    top: 7,
  },
  Reflow: {
    ...defaultStyle,
    background: colors.BLUE_50,
    top: 7,
  },
  DispatchSynthMouseMove: {
    ...defaultStyle,
    background: colors.ORANGE_70,
    top: 8,
  },
  DisplayList: {
    ...defaultStyle,
    background: colors.PURPLE_50,
    top: 9,
  },
  LayerBuilding: {
    ...defaultStyle,
    background: colors.ORANGE_50,
    top: 9,
  },
  Rasterize: {
    ...defaultStyle,
    background: colors.GREEN_50,
    top: 10,
  },
  ForwardTransaction: {
    ...defaultStyle,
    background: colors.RED_70,
    top: 11,
  },
  NotifyDidPaint: {
    ...defaultStyle,
    background: colors.GREY_40,
    top: 12,
  },
  LayerTransaction: {
    ...defaultStyle,
    background: colors.RED_70,
  },
  Composite: {
    ...defaultStyle,
    background: colors.BLUE_50,
  },
  Vsync: {
    ...defaultStyle,
    background: 'rgb(255, 128, 0)',
  },
  LayerContentGPU: {
    ...defaultStyle,
    background: 'rgba(0,200,0,0.5)',
  },
  LayerCompositorGPU: {
    ...defaultStyle,
    background: 'rgba(0,200,0,0.5)',
  },
  LayerOther: {
    ...defaultStyle,
    background: 'rgb(200,0,0)',
  },
  Jank: {
    ...defaultStyle,
    background: 'hsl(347, 100%, 60%)',
    borderLeft: colors.RED_50,
    borderRight: colors.RED_50,
    squareCorners: true,
  },
  // BHR markers are displayed in the timeline only if jank markers are
  // unavailable. Let's style them like Jank markers.
  'BHR-detected hang': {
    ...defaultStyle,
    background: 'hsl(347, 100%, 60%)',
    borderLeft: colors.RED_50,
    borderRight: colors.RED_50,
    squareCorners: true,
  },
  // Memory:
  GCMajor: {
    ...gcStyle,
    squareCorners: true,
    top: 0,
  },
  GCSlice: gcStyle,
  GCMinor: gcStyle,
  'GC Interrupt': gcStyle,
  CC: {
    ...ccStyle,
    squareCorners: true,
    top: 0,
  },
  CCSlice: ccStyle,
  ForgetSkippable: ccStyle,
  // Note: these Idle* markers have been removed in Firefox 99 (Bug 1752646),
  // but they're still here for compatibility for older profiles.
  IdleCCSlice: ccStyle,
  IdleForgetSkippable: ccStyle,

  // IO:
  FileIO: {
    ...defaultStyle,
    background: colors.BLUE_50,
  },

  IPCOut: {
    ...defaultStyle,
    background: colors.BLUE_50,
    top: 2,
  },
  SyncIPCOut: {
    ...defaultStyle,
    background: colors.BLUE_70,
    top: 6,
  },
  IPCIn: {
    ...defaultStyle,
    background: colors.PURPLE_40,
    top: 13,
  },
  SyncIPCIn: {
    ...defaultStyle,
    background: colors.PURPLE_70,
    top: 17,
  },
};

export const overlayFills = {
  HOVERED: 'hsla(0,0%,100%,0.3)',
  PRESSED: 'rgba(0,0,0,0.3)',
};
