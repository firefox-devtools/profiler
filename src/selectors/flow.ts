/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import { createSelector } from 'reselect';

import {
  computeProfileFlowInfo,
  computeFlowTiming,
} from '../profile-logic/marker-data';
import { getThreadSelectors } from './per-thread';
import { getActiveFlows } from './url-state';
import type { ThreadSelectors } from './per-thread';
import {
  getThreads,
  getMarkerSchema,
  getRawProfileSharedData,
} from './profile';

import type {
  Selector,
  State,
  MarkerIndex,
  Marker,
  ProfileFlowInfo,
  FlowTiming,
} from 'firefox-profiler/types';

function _arraysShallowEqual(arr1: any[], arr2: any[]): boolean {
  return arr1.length === arr2.length && arr1.every((val, i) => val === arr2[i]);
}

function _createSelectorForAllThreads<T>(
  f: (selectors: ThreadSelectors, state: State) => T
): Selector<T[]> {
  let previousOutputPerThread: T[] = [];
  return function recomputeSelectorForAllThreads(state: State): T[] {
    const threads = getThreads(state);
    let outputPerThread = threads.map((_thread, i) => {
      const threadSelectors = getThreadSelectors(i);
      return f(threadSelectors, state);
    });
    if (_arraysShallowEqual(outputPerThread, previousOutputPerThread)) {
      outputPerThread = previousOutputPerThread;
    }
    previousOutputPerThread = outputPerThread;
    return outputPerThread;
  };
}

export const getFullMarkerListPerThread: Selector<Marker[][]> =
  _createSelectorForAllThreads(({ getFullMarkerList }, state) =>
    getFullMarkerList(state)
  );

export const getMarkerChartLabelGetterPerThread: Selector<
  Array<(marker: MarkerIndex) => string>
> = _createSelectorForAllThreads(({ getMarkerChartLabelGetter }, state) =>
  getMarkerChartLabelGetter(state)
);

export const getProfileFlowInfo: Selector<ProfileFlowInfo> = createSelector(
  getFullMarkerListPerThread,
  getThreads,
  getMarkerSchema,
  getRawProfileSharedData,
  computeProfileFlowInfo
);

export const getFlowTiming: Selector<FlowTiming> = createSelector(
  getProfileFlowInfo,
  getActiveFlows,
  computeFlowTiming
);
