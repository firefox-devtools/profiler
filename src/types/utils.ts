/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

/**
 * This type serves as documentation for how an array is meant to be used, but does
 * not support type checking. We often use an Array instead of a Map to translate
 * one type of index into another type of index. This is similar to how we use the
 * Map<K,V> type, but with the Array.
 */
export type IndexedArray<_IndexType, Value> = Array<Value>;

export type MixedObject = { [key: string]: unknown };
