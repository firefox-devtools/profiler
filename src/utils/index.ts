/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import type {
  Marker,
  Milliseconds,
  StartEndRange,
  KeyboardModifiers,
} from 'firefox-profiler/types';
import type {
  MouseEvent as SyntheticMouseEvent,
  KeyboardEvent as SyntheticKeyboardEvent,
} from 'react';

/**
 * Firefox has issues switching quickly between fill style colors, as the CSS color
 * is fully parsed each time it is set. As a mitigation, provide a class that only
 * switches the color when it's really needed.
 */
export class FastFillStyle {
  _ctx: CanvasRenderingContext2D;
  _previousFillColor: string;

  constructor(ctx: CanvasRenderingContext2D) {
    this._ctx = ctx;
    this._previousFillColor = '';
  }

  set(fillStyle: string) {
    if (fillStyle !== this._previousFillColor) {
      this._ctx.fillStyle = fillStyle;
      this._previousFillColor = fillStyle;
    }
  }
}

/**
 * Perform a simple shallow object equality check.
 */
export function objectShallowEquals<
  A extends Record<string, any>,
  B extends Record<string, any>,
>(a: A, b: B): boolean {
  let aLength = 0;
  let bLength = 0;
  for (const key in a) {
    if (Object.prototype.hasOwnProperty.call(a, key)) {
      aLength++;
      if ((a as any)[key] !== (b as any)[key]) {
        return false;
      }
    }
  }
  for (const key in b) {
    if (Object.prototype.hasOwnProperty.call(b, key)) {
      bLength++;
    }
  }
  return aLength === bLength;
}

/**
 * Don't completely trust URLs that get displayed to the screen. These could
 * be extraordinarily long data-uris. This function could also be updated to
 * return a nicely formatted React span.
 */
export function displayNiceUrl(rawUrl: string): string {
  if (rawUrl.length < 200) {
    return rawUrl;
  }
  return `${rawUrl.slice(0, 150)}...${rawUrl.slice(-20)}`;
}

/**
 * This function handles the logic of converting both Interval and Instant markers
 * into a range selection. For instant markers, it creates a range that is 5%
 * of the range selection (typically the committed range).
 */
export function getStartEndRangeForMarker(
  rangeStart: Milliseconds,
  rangeEnd: Milliseconds,
  marker: Marker
): StartEndRange {
  let start = marker.start;
  let end = marker.end;

  if (end === null) {
    const delta = (rangeEnd - rangeStart) / 40;
    start = marker.start - delta;
    end = marker.start + delta;
  }

  return { start, end };
}

/**
 * This logic is shared between multiple components, but it is used to determine how a
 * track gets selected, based on the modifiers used.
 */
export function getTrackSelectionModifiers(
  event:
    | MouseEvent
    | KeyboardEvent
    | SyntheticMouseEvent<any>
    | SyntheticKeyboardEvent<any>
): KeyboardModifiers {
  return {
    ctrlOrMeta: (event.ctrlKey || event.metaKey) && !event.altKey,
    shift: event.shiftKey && !event.altKey,
  };
}

export function countPositiveValues(arr: Array<number>): number {
  let count = 0;
  for (let i = 0; i < arr.length; i++) {
    if (arr[i] > 0) {
      count++;
    }
  }
  return count;
}

/**
 * Find the highest number among the map values and return its key.
 * Returns undefined if the map is empty.
 * If multiple entries with the highest value exist, it returns the key of the
 * first encountered highest value.
 */
export function mapGetKeyWithMaxValue<K>(map: Map<K, number>): K | undefined {
  let maxValue = -Infinity;
  let keyForMaxValue;
  for (const [key, value] of map) {
    if (value > maxValue) {
      maxValue = value;
      keyForMaxValue = key;
    }
  }
  return keyForMaxValue;
}
