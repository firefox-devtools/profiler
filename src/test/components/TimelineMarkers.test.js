/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import * as React from 'react';
import { Provider } from 'react-redux';

// This module is mocked.
import copy from 'copy-to-clipboard';

import {
  render,
  screen,
  fireEvent,
  act,
} from 'firefox-profiler/test/fixtures/testing-library';
import {
  TimelineMarkersOverview,
  TimelineMarkersMemory,
  MIN_MARKER_WIDTH,
} from '../../components/timeline/Markers';
import { MaybeMarkerContextMenu } from '../../components/shared/MarkerContextMenu';
import { overlayFills } from '../../profile-logic/marker-styles';
import { ensureExists } from '../../utils/flow';

import {
  autoMockCanvasContext,
  flushDrawLog,
} from '../fixtures/mocks/canvas-context';
import { storeWithProfile } from '../fixtures/stores';
import { getProfileWithMarkers } from '../fixtures/profiles/processed-profile';
import {
  getMouseEvent,
  addRootOverlayElement,
  removeRootOverlayElement,
  fireFullClick,
  fireFullContextMenu,
} from '../fixtures/utils';
import { mockRaf } from '../fixtures/mocks/request-animation-frame';
import {
  autoMockElementSize,
  setMockedElementSize,
} from '../fixtures/mocks/element-size';
import {
  autoMockIntersectionObserver,
  triggerIntersectionObservers,
} from '../fixtures/mocks/intersection-observer';
import { triggerResizeObservers } from '../fixtures/mocks/resize-observer';

import type { CssPixels } from 'firefox-profiler/types';

const GRAPH_WIDTH = 200;
const GRAPH_HEIGHT = 300;

function setupWithMarkers(
  {
    rangeStart,
    rangeEnd,
    component,
  }: {
    rangeStart: number,
    rangeEnd: number,
    component?: typeof TimelineMarkersOverview,
  },
  ...markersPerThread
) {
  const flushRafCalls = mockRaf();

  const profile = getProfileWithMarkers(...markersPerThread);
  const TimelineMarkersComponent = component ?? TimelineMarkersOverview;

  const renderResult = render(
    <Provider store={storeWithProfile(profile)}>
      <>
        <TimelineMarkersComponent
          rangeStart={rangeStart}
          rangeEnd={rangeEnd}
          threadsKey={0}
          onSelect={() => {}}
        />
        <MaybeMarkerContextMenu />
      </>
    </Provider>
  );

  flushRafCalls();

  function getContextMenu() {
    return ensureExists(
      document.querySelector('.react-contextmenu'),
      `Couldn't find the context menu.`
    );
  }

  function tryToGetMarkerTooltip() {
    return document.querySelector('.tooltipMarker');
  }

  function clickOnMenuItem(stringOrRegexp) {
    fireFullClick(screen.getByText(stringOrRegexp));
  }

  function fireMouseEvent(eventName, options) {
    fireEvent(
      ensureExists(
        document.querySelector('canvas'),
        `Couldn't find the canvas element`
      ),
      getMouseEvent(eventName, options)
    );
  }

  function getPositioningOptions({ x, y }) {
    // These positioning options will be sent to all our mouse events. Note
    // that the values aren't really consistent, especially offsetY and
    // pageY shouldn't be the same, but in the context of our test this will
    // be good enough.
    // pageX/Y values control the position of the tooltip so it's not super
    // important.
    // offsetX/Y are more important as they're used to find which node is
    // actually clicked.
    const positioningOptions = {
      offsetX: x,
      offsetY: y,
      clientX: x,
      clientY: y,
      pageX: x,
      pageY: y,
    };

    return positioningOptions;
  }

  // Note to a future developer: the x/y values can be derived from the
  // array returned by ctx.__flushDrawLog().
  function rightClick(where: { x: CssPixels, y: CssPixels }) {
    const positioningOptions = getPositioningOptions(where);
    const canvas = ensureExists(
      document.querySelector('canvas'),
      `Couldn't find the canvas element`
    );

    // Because different components listen to different events, we trigger
    // all the right events, to be as close as possible to the real stuff.
    fireMouseEvent('mousemove', positioningOptions);
    fireFullContextMenu(canvas, positioningOptions);

    flushRafCalls();
  }

  function mouseOver(where: { x: CssPixels, y: CssPixels }) {
    const positioningOptions = getPositioningOptions(where);
    fireMouseEvent('mousemove', positioningOptions);
    flushRafCalls();
  }

  return {
    rightClick,
    mouseOver,
    getContextMenu,
    tryToGetMarkerTooltip,
    clickOnMenuItem,
    flushRafCalls,
    ...renderResult,
  };
}

describe('TimelineMarkers', function () {
  autoMockCanvasContext();
  autoMockElementSize({ width: GRAPH_WIDTH, height: GRAPH_HEIGHT });
  autoMockIntersectionObserver();
  // We will be hovering over element with a tooltip. It requires root overlay
  // element to be present in DOM.
  beforeEach(addRootOverlayElement);
  afterEach(removeRootOverlayElement);

  it('renders correctly overview markers', () => {
    window.devicePixelRatio = 1;

    const { container } = setupWithMarkers({ rangeStart: 0, rangeEnd: 15 }, [
      ['DOMEvent', 0, 10, { type: 'tracing', category: 'JS' }],
      ['DOMEvent', 0, 10, { type: 'tracing', category: 'JS' }],
      ['DOMEvent', 5, 15, { type: 'DOMEvent', latency: 3.7, eventType: 'msg' }],
      ['Paint', 2, 13, { type: 'tracing', category: 'Paint' }],
      ['Navigation', 2, 6, { type: 'tracing', category: 'Navigation' }],
      ['Layout', 6, 8, { type: 'tracing', category: 'Layout' }],
      // These 2 will be ignored.
      ['CC', 0, 5],
      ['GCMajor', 5, 10],
    ]);

    const drawCalls = flushDrawLog();

    expect(container.firstChild).toMatchSnapshot();
    expect(drawCalls).toMatchSnapshot();

    delete window.devicePixelRatio;
  });

  it('renders correctly memory markers', () => {
    window.devicePixelRatio = 1;

    const { container } = setupWithMarkers(
      { rangeStart: 0, rangeEnd: 15, component: TimelineMarkersMemory },
      [
        // The first one will be ignored.
        ['DOMEvent', 0, 10, { type: 'tracing', category: 'JS' }],
        ['CC', 0, 5],
        ['GCMajor', 5, 10],
      ]
    );

    const drawCalls = flushDrawLog();

    expect(container.firstChild).toMatchSnapshot();
    expect(drawCalls).toMatchSnapshot();

    delete window.devicePixelRatio;
  });

  it('does not render several dot markers in the same position', () => {
    window.devicePixelRatio = 2;

    setupWithMarkers({ rangeStart: 0, rangeEnd: 15000 }, [
      // 2 very close dot markers. They shouldn't be drawn both together.
      ['DOMEvent', 5000],
      ['DOMEvent', 5001],
      // This is a longer marker starting at the same place, it should always be drawn
      ['DOMEvent', 5001, 7000],
    ]);

    const drawCalls = flushDrawLog();

    // We filter on height to get only 1 relevant fillRect operation for each marker.
    const fillRectOperations = drawCalls.filter(
      ([operation, , , , height]) => operation === 'fillRect' && height > 1
    );

    // Here 2 markers should be drawn: the first dot, and the long marker.
    expect(fillRectOperations).toHaveLength(2);
    expect(
      fillRectOperations.every(
        ([, , , width]) => width >= MIN_MARKER_WIDTH / window.devicePixelRatio
      )
    ).toBe(true);

    delete window.devicePixelRatio;
  });

  it('renders the first marker tooltip when hovered', () => {
    const { tryToGetMarkerTooltip, mouseOver } = setupWithMarkers(
      { rangeStart: 0, rangeEnd: 10 },
      [
        ['DOMEvent', 0, 3],
        ['Load', 6, 10, { type: 'tracing', category: 'Navigation' }],
      ]
    );

    // Tooltip should not be visible yet.
    expect(tryToGetMarkerTooltip()).not.toBeInTheDocument();

    // The "DOMEvent" marker is drawn from 0,0 to 5,60.
    mouseOver({ x: 30, y: 2 });

    // It should be visible after hovering the first marker.
    expect(tryToGetMarkerTooltip()).toBeInTheDocument();
  });

  describe('displays context menus', () => {
    beforeEach(() => {
      // Always use fake timers when dealing with context menus.
      jest.useFakeTimers();
    });

    it('when right clicking on a marker', () => {
      const { rightClick, getContextMenu, clickOnMenuItem } = setupWithMarkers(
        { rangeStart: 0, rangeEnd: 10 },
        [['DOMEvent', 0, 10, { type: 'DOMEvent', eventType: 'mousedown' }]]
      );

      // The "DOMEvent" marker is drawn from 0,0 to 5,200.
      rightClick({ x: 50, y: 2 });

      act(() => jest.runAllTimers());

      expect(getContextMenu()).toHaveClass('react-contextmenu--visible');

      clickOnMenuItem('Copy description');
      expect(copy).toHaveBeenLastCalledWith('DOMEvent â€” mousedown');
      expect(getContextMenu()).not.toHaveClass('react-contextmenu--visible');

      act(() => jest.runAllTimers());

      expect(document.querySelector('react-contextmenu')).toBeFalsy();
    });

    it('when right clicking on markers in a sequence', () => {
      const { rightClick, getContextMenu, clickOnMenuItem } = setupWithMarkers(
        { rangeStart: 0, rangeEnd: 10 },
        [
          ['DOMEvent', 0, 3],
          ['Load', 6, 10, { type: 'tracing', category: 'Navigation' }],
        ]
      );

      // The "DOMEvent" marker is drawn from 0,0 to 5,60.
      rightClick({ x: 30, y: 2 });
      act(() => jest.runAllTimers());

      // The "Navigation" marker is drawn from 0,120 to 5,200.
      rightClick({ x: 160, y: 2 });
      act(() => jest.runAllTimers());

      expect(getContextMenu()).toHaveClass('react-contextmenu--visible');

      clickOnMenuItem('Copy description');
      expect(copy).toHaveBeenLastCalledWith('Load');

      expect(getContextMenu()).not.toHaveClass('react-contextmenu--visible');

      act(() => jest.runAllTimers());
      expect(document.querySelector('react-contextmenu')).toBeFalsy();
    });

    it('and still highlights other markers when hovering them', () => {
      const { rightClick, mouseOver, getContextMenu } = setupWithMarkers(
        { rangeStart: 0, rangeEnd: 10 },
        [
          ['DOMEvent', 0, 3],
          ['DOMEvent', 6, 10],
        ]
      );

      // The "DOMEvent" marker is drawn from 0,0 to 5,60.
      rightClick({ x: 30, y: 2 });
      expect(getContextMenu()).toHaveClass('react-contextmenu--visible');

      flushDrawLog();
      // The "Marker B" marker is drawn from 0,120 to 5,200.
      mouseOver({ x: 160, y: 1 });

      const drawCalls = flushDrawLog();

      // Expect that we have one marker with hovered color
      const callsWithHoveredColor = drawCalls.filter(
        ([, argument]) => argument === overlayFills.HOVERED
      );
      expect(callsWithHoveredColor).toHaveLength(1);
    });
  });
});

describe('TimelineMarkers with intersection observer', function () {
  autoMockCanvasContext();
  autoMockElementSize({ width: GRAPH_WIDTH, height: GRAPH_HEIGHT });
  autoMockIntersectionObserver(false);

  function setup() {
    const setupResults = setupWithMarkers({ rangeStart: 0, rangeEnd: 15 }, [
      ['DOMEvent', 0, 10],
      ['DOMEvent', 0, 10],
    ]);

    /**
     * Coordinate the flushing of the requestAnimationFrame and the draw calls.
     */
    function getContextDrawCalls() {
      setupResults.flushRafCalls();
      return flushDrawLog();
    }

    return { ...setupResults, getContextDrawCalls };
  }

  it('will not draw before the intersection observer', () => {
    const { getContextDrawCalls } = setup();
    const drawCalls = getContextDrawCalls();
    // Make sure that marker graph is not drawn yet.
    expect(drawCalls.some(([operation]) => operation === 'fillRect')).toBe(
      false
    );
  });

  it('will not draw after the intersection observer if it is not intersecting', () => {
    const { getContextDrawCalls } = setup();
    let drawCalls = getContextDrawCalls();

    // Make sure that marker graph is not drawn yet.
    expect(drawCalls.some(([operation]) => operation === 'beginPath')).toBe(
      false
    );

    // Now let's trigger the intersection observer and make sure that it still
    // doesn't draw it.
    triggerIntersectionObservers({ isIntersecting: false });
    drawCalls = getContextDrawCalls();
    expect(drawCalls.some(([operation]) => operation === 'fillRect')).toBe(
      false
    );
  });

  it('will draw after the intersection observer if it is intersecting', () => {
    const { getContextDrawCalls } = setup();
    let drawCalls = getContextDrawCalls();

    // Make sure that marker graph is not drawn yet.
    expect(drawCalls.some(([operation]) => operation === 'fillRect')).toBe(
      false
    );

    // Now let's trigger the intersection observer and make sure that it draws it.
    triggerIntersectionObservers({ isIntersecting: true });
    drawCalls = getContextDrawCalls();
    expect(drawCalls.some(([operation]) => operation === 'fillRect')).toBe(
      true
    );
  });

  it('will redraw after it becomes visible again', () => {
    const { getContextDrawCalls } = setup();
    let drawCalls = getContextDrawCalls();

    // Make sure that marker graph is not drawn yet.
    expect(drawCalls.some(([operation]) => operation === 'fillRect')).toBe(
      false
    );

    // Now let's trigger the intersection observer and make sure that it draws it.
    triggerIntersectionObservers({ isIntersecting: true });
    drawCalls = getContextDrawCalls();
    expect(drawCalls.some(([operation]) => operation === 'fillRect')).toBe(
      true
    );

    // Now it goes out of view again. Make sure that we don't redraw.
    triggerIntersectionObservers({ isIntersecting: false });
    drawCalls = getContextDrawCalls();
    expect(drawCalls.some(([operation]) => operation === 'fillRect')).toBe(
      false
    );

    // Send out the resize with a width change.
    // By changing the "fake" result of getBoundingClientRect, we ensure that
    // the pure components rerender because their `width` props change.
    setMockedElementSize({ width: GRAPH_WIDTH * 2, height: GRAPH_HEIGHT });
    triggerResizeObservers();
    drawCalls = getContextDrawCalls();
    // It should still be not drawn yet.
    expect(drawCalls.some(([operation]) => operation === 'fillRect')).toBe(
      false
    );

    // Now let's trigger the intersection observer again and make sure that it redraws.
    triggerIntersectionObservers({ isIntersecting: true });
    drawCalls = getContextDrawCalls();
    expect(drawCalls.some(([operation]) => operation === 'fillRect')).toBe(
      true
    );
  });
});
