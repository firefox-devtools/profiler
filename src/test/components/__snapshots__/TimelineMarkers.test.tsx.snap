// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`TimelineMarkers renders correctly memory markers 1`] = `
<div
  class="timelineMarkers timelineMarkersMemory selected"
  data-testid="TimelineMarkersMemory"
>
  <div
    class="react-contextmenu-wrapper"
  >
    <div>
      <canvas
        class="timelineMarkersCanvas"
        height="300"
        width="200"
      />
    </div>
  </div>
</div>
`;

exports[`TimelineMarkers renders correctly memory markers 2`] = `
Array [
  Array [
    "clearRect",
    0,
    0,
    200,
    300,
  ],
  Array [
    "scale",
    1,
    1,
  ],
  Array [
    "set fillStyle",
    "#ffc600",
  ],
  Array [
    "fillRect",
    0,
    0,
    66.66666666666666,
    6,
  ],
  Array [
    "set fillStyle",
    "#ff9400",
  ],
  Array [
    "fillRect",
    67,
    0,
    66.66666666666666,
    6,
  ],
  Array [
    "scale",
    1,
    1,
  ],
]
`;

exports[`TimelineMarkers renders correctly overview markers 1`] = `
<div
  class="timelineMarkers selected"
  data-testid="TimelineMarkersOverview"
>
  <div
    class="react-contextmenu-wrapper"
  >
    <div>
      <canvas
        class="timelineMarkersCanvas"
        height="300"
        width="200"
      />
    </div>
  </div>
</div>
`;

exports[`TimelineMarkers renders correctly overview markers 2`] = `
Array [
  Array [
    "clearRect",
    0,
    0,
    200,
    300,
  ],
  Array [
    "scale",
    1,
    1,
  ],
  Array [
    "set fillStyle",
    "black",
  ],
  Array [
    "fillRect",
    1,
    0,
    131.33333333333331,
    1,
  ],
  Array [
    "fillRect",
    0,
    1,
    133.33333333333331,
    4,
  ],
  Array [
    "fillRect",
    1,
    5,
    131.33333333333331,
    1,
  ],
  Array [
    "set fillStyle",
    "black",
  ],
  Array [
    "fillRect",
    1,
    0,
    131.33333333333331,
    1,
  ],
  Array [
    "fillRect",
    0,
    1,
    133.33333333333331,
    4,
  ],
  Array [
    "fillRect",
    1,
    5,
    131.33333333333331,
    1,
  ],
  Array [
    "set fillStyle",
    "black",
  ],
  Array [
    "fillRect",
    28,
    0,
    144.66666666666666,
    1,
  ],
  Array [
    "fillRect",
    27,
    1,
    146.66666666666666,
    4,
  ],
  Array [
    "fillRect",
    28,
    5,
    144.66666666666666,
    1,
  ],
  Array [
    "set fillStyle",
    "black",
  ],
  Array [
    "fillRect",
    28,
    0,
    51.333333333333336,
    1,
  ],
  Array [
    "fillRect",
    27,
    1,
    53.333333333333336,
    4,
  ],
  Array [
    "fillRect",
    28,
    5,
    51.333333333333336,
    1,
  ],
  Array [
    "set fillStyle",
    "black",
  ],
  Array [
    "fillRect",
    68,
    0,
    131.33333333333331,
    1,
  ],
  Array [
    "fillRect",
    67,
    1,
    133.33333333333331,
    4,
  ],
  Array [
    "fillRect",
    68,
    5,
    131.33333333333331,
    1,
  ],
  Array [
    "set fillStyle",
    "black",
  ],
  Array [
    "fillRect",
    81,
    0,
    24.666666666666668,
    1,
  ],
  Array [
    "fillRect",
    80,
    1,
    26.666666666666668,
    4,
  ],
  Array [
    "fillRect",
    81,
    5,
    24.666666666666668,
    1,
  ],
  Array [
    "scale",
    1,
    1,
  ],
]
`;
